### ISSUE TITLE
`flutter upgrade` fails on Windows/ARM because of locked cache folder

### ISSUE THREAD

**Issue by timsneath** [flutter/flutter#114753]
## Steps to Reproduce

Running Flutter on Windows ARM, `flutter upgrade` doesn't work because the cache directory gets locked during the install process. 

## Logs
This is from a freshly-rebooted machine which has not run VSCode, Flutter or Dart since the reboot.

<details>

```
Checking Dart SDK version...
Downloading Dart SDK from Flutter engine 0886c7d0b169690a05d4c39f9d6ae59200081756...
Rename-Item: C:\src\flutter\bin\internal\update_dart_sdk.ps1:52
Line |
  52 |      Rename-Item $dartSdkPath "$oldDartSdkPrefix$oldDartSdkSuffix"
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Access to the path 'C:\src\flutter\bin\cache\dart-sdk' is denied.

Error: Unable to update Dart SDK. Retrying...
ERROR: Input redirection is not supported, exiting the process immediately.
Checking Dart SDK version...
Downloading Dart SDK from Flutter engine 0886c7d0b169690a05d4c39f9d6ae59200081756...
Terminate batch job (Y/N)? y

[C:\src\win32] rd /s \src\flutter\bin\cache
\src\flutter\bin\cache, Are you sure (Y/N)? y

[C:\src\win32] flutter upgrade
Checking Dart SDK version...
Downloading Dart SDK from Flutter engine ...
Expanding downloaded archive...
Building flutter tool...
Running pub upgrade...
... proceeds to complete successfully
```

</details>

## Expected results

Flutter should obviously be able to delete the folder. For some reason, rename fails but delete succeeds. I wonder if we need a fallback option where if the rename fails, we just blow the cache away?

**Comment by danagbemava-nc** [comment]
Labeling this for further investigation as I'm unable to verify this locally at the moment.

**Comment by christopherfujino** [comment]
@loic-sharma or @cbracken, can one of you verify this on a real WIndows ARM machine? I wonder if this is related to:

1. The Windows ARM build
2. Running in parallels' virtualized windows filesystem
3. something else

**Comment by timsneath** [comment]
FWIW, I'm able to reproduce it on a Volterra dev kit:

```txt
C:\src\flutter>bin\flutter upgrade
Checking Dart SDK version...
Downloading Dart SDK from Flutter engine ...
Expanding downloaded archive...
Building flutter tool...
Running pub upgrade...
Upgrading Flutter to 3.5.0-12.0.pre.165 from 3.5.0-12.0.pre.101 in C:\src\flutter...
Checking Dart SDK version...
Downloading Dart SDK from Flutter engine 004a305166d57756cf7c5f6255dd09a5cf43479d...
Rename-Item: C:\src\flutter\bin\internal\update_dart_sdk.ps1:52
Line |
  52 |      Rename-Item $dartSdkPath "$oldDartSdkPrefix$oldDartSdkSuffix"
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Access to the path 'C:\src\flutter\bin\cache\dart-sdk' is denied.
```

**Comment by christopherfujino** [comment]
> FWIW, I'm able to reproduce it on a [Volterra dev kit](https://blogs.windows.com/windowsdeveloper/2022/10/24/available-today-windows-dev-kit-2023-aka-project-volterra/), which seems to rule out # 2 from the above.
> 
> ```
> C:\src\flutter>bin\flutter upgrade
> Checking Dart SDK version...
> Downloading Dart SDK from Flutter engine ...
> Expanding downloaded archive...
> Building flutter tool...
> Running pub upgrade...
> Upgrading Flutter to 3.5.0-12.0.pre.165 from 3.5.0-12.0.pre.101 in C:\src\flutter...
> Checking Dart SDK version...
> Downloading Dart SDK from Flutter engine 004a305166d57756cf7c5f6255dd09a5cf43479d...
> Rename-Item: C:\src\flutter\bin\internal\update_dart_sdk.ps1:52
> Line |
>   52 |      Rename-Item $dartSdkPath "$oldDartSdkPrefix$oldDartSdkSuffix"
>      |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>      | Access to the path 'C:\src\flutter\bin\cache\dart-sdk' is denied.
> ```
> 
> Machine is left in this state; let me know if there are any diagnostics you'd like me to run?

Oh wow, very interesting. And I can't even think of any diagnostics to run.

**Comment by ngaretou** [comment]
FWIW messing around with this a bit `flutter doctor` of all things somehow triggered the upgrade...

**Comment by RoarGronmo** [comment]
FYI: This happens on all flavors of flutter, master/main, beta and stable.

Since it is at least one year since someone commented here, are any working on it?

**Comment by roylan34** [comment]
I encountered this issue. What i did is to disconnected the running app from emulator or device before executing `flutter upgrade`.

**Comment by RoarGronmo** [comment]
For now I never upgrade or changes channel when AS is started.

I upgrade/change channel from my flutter program folder in powershell. It doesn't fail when using this method.

**Comment by sharmaruchi30** [comment]
I had the same issue just now, and a system restart worked for me.

**Comment by matt3o** [comment]
**Close all open flutter applications**, naturally they might block flutter from upgrading. Fixed it for me.

@ flutter Team a short message to close all open instances of flutter applications would be great in this case, I ran into the same issue.

**Comment by cbracken** [comment]
This seems like a pretty reasonable idea to me. Thoughts @loic-sharma (for windows) or @andrewkolos (for tool)?

**Comment by loic-sharma** [comment]
That seems reasonable mitigation. I suspect we could wrap this `Rename-Item` with a `Try` `Catch`:

https://github.com/flutter/flutter/blob/945067ae917c26f294417efd5894a09319650ede/bin/internal/update_dart_sdk.ps1#L81

If it fails, we could output an error message that recommends the user restarts their computer. I'd be happy to review such a patch, feel free to ping me if anyone creates a pull request :)

**Comment by KaelTeck** [comment]
I've solved this issue by killing adb and dart process in the task manager

**Comment by deakjahn** [comment]
If Android Studio is running and the upgrade is started from the CLI, the error is as above. But even then this is a regression because it used to work earlier. At any rate, the script should immediately stop with a reasonable message and no to enter an infinite loop.

Second, the same Flutter Upgrade from the Flutter menu of Android Studio doesn't work, either (it obviously starts the external script while AS itself is running, so the issue is the same). Again, this used to work earlier so it's not really an user error. If the IDE has a feature like that, that feature should work.

**Comment by esDotDev** [comment]
Yes, this has worked for 5 years, we shouldn't need to close all flutter instances to do a simple upgrade from the IDE terminal.

**Comment by krll-kov** [comment]
The process holding `bin\cache\dart-sdk` is the Dart VM of the first upgrade phase, still alive while the second phase re-runs `update_dart_sdk.ps1`, and only when that Dart SDK is the x64 one under emulation.

Two runs of `flutter upgrade --force`, 3.47.4 to 3.47.5, Windows 11 ARM64, differing only in the architecture of the cached Dart SDK:

- x64, which is what the official archive ships (`releases_windows.json` reports `dart_sdk_arch: x64`): `Rename-Item` fails with "Access to the path ...\bin\cache\dart-sdk is denied" on all three retries and the upgrade exits 1.
- arm64, after deleting `bin\cache\dart-sdk` and its stamp and running flutter once with `FLUTTER_HOST_ARCH=arm64`: the same upgrade completes in one go, exit 0.

So on ARM64 this hits every install from the official zip, whatever else is running.

Separately, the delay between retries never happens: `shared.bat:151` uses `timeout /t 5 /nobreak` without the `2>NUL` that line 186 has, and with stdin redirected timeout exits immediately
